File: ios/iOSBaseProject/Domain/UseCases/Auth/SignUpUseCase.swift
//
//  SignUpUseCase.swift
//  BaatCheet
//
//  Domain layer - Sign Up Use Case
//

import Foundation

/// Use case for registering a new user.
/// Validates credentials and delegates to the repository.
protocol SignUpUseCaseProtocol {
    func execute(credentials: AuthCredentials) async -> AuthResult
}

final class SignUpUseCase: SignUpUseCaseProtocol {
    private let repository: AuthRepositoryProtocol
    
    init(repository: AuthRepositoryProtocol) {
        self.repository = repository
    }
    
    func execute(credentials: AuthCredentials) async -> AuthResult {
        // Validate credentials
        guard !credentials.email.isEmpty else {
            return .failure(.invalidEmail)
        }
        
        guard credentials.isEmailValid else {
            return .failure(.invalidEmail)
        }
        
        guard !credentials.password.isEmpty else {
            return .failure(.weakPassword)
        }
        
        guard credentials.isPasswordValid else {
            return .failure(.weakPassword)
        }
        
        // Delegate to repository
        return await repository.signUp(credentials: credentials)
    }
}

File: android/app/src/main/java/com/baatcheet/app/domain/usecase/auth/SignUpUseCase.kt
package com.baatcheet.app.domain.usecase.auth

import com.baatcheet.app.domain.model.AuthResult
import com.baatcheet.app.domain.repository.AuthRepository
import javax.inject.Inject

/**
 * Sign Up Use Case
 * 
 * Handles the business logic for user registration.
 * Validates inputs and delegates to the repository.
 */
class SignUpUseCase @Inject constructor(
    private val authRepository: AuthRepository
) {
    /**
     * Execute sign up with email and password
     */
    suspend operator fun invoke(
        email: String,
        password: String,
        firstName: String? = null,
        lastName: String? = null
    ): AuthResult {
        // Validate inputs
        if (email.isBlank()) {
            return AuthResult.Failure(IllegalArgumentException("Email cannot be empty"))
        }
        if (password.isBlank()) {
            return AuthResult.Failure(IllegalArgumentException("Password cannot be empty"))
        }
        if (password.length < 8) {
            return AuthResult.Failure(IllegalArgumentException("Password must be at least 8 characters"))
        }
        
        return authRepository.signUp(email, password, firstName, lastName)
    }
}
